--- a/drivers/staging/vc04_services/bcm2835-audio/bcm2835.c	2025-06-24 10:48:26.000000000 +0200
+++ b/drivers/staging/vc04_services/bcm2835-audio/bcm2835.c	2025-11-13 23:38:33.034297756 +0100
@@ -11,7 +11,7 @@
 #include "bcm2835.h"
 #include <soc/bcm2835/raspberrypi-firmware.h>
 
-static bool enable_hdmi, enable_hdmi0, enable_hdmi1;
+static bool enable_hdmi = true, enable_hdmi0, enable_hdmi1;
 static bool enable_headphones = true;
 static int num_channels = MAX_SUBSTREAMS;
 
