diff --git a/arch/riscv/boot/dts/starfive/jh7110-starfive-visionfive-2.dtsi b/arch/riscv/boot/dts/starfive/jh7110-starfive-visionfive-2.dtsi
index 752bb0b6fd00..93670da6cabd 100644
--- a/arch/riscv/boot/dts/starfive/jh7110-starfive-visionfive-2.dtsi
+++ b/arch/riscv/boot/dts/starfive/jh7110-starfive-visionfive-2.dtsi
@@ -30,7 +30,7 @@ cpus {
 
 	memory@40000000 {
 		device_type = "memory";
-		reg = <0x0 0x40000000 0x1 0x0>;
+		reg = <0x0 0x40000000 0x2 0x0>;
 	};
 
 	thermal-zones {
