From 313309c07e904ba48386568c361f4a3265829a81 Mon Sep 17 00:00:00 2001
From: Ganga Ram <Ganga.Ram@tii.ae>
Date: Tue, 10 Oct 2023 17:07:49 +0400
Subject: [PATCH] Boot environment for Microchip Iciclle Kit

Signed-off-by: Ganga Ram <Ganga.Ram@tii.ae>
---
 include/configs/microchip_mpfs_icicle.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/configs/microchip_mpfs_icicle.h b/include/configs/microchip_mpfs_icicle.h
index 0b9eb59bc4..59b4465f1e 100644
--- a/include/configs/microchip_mpfs_icicle.h
+++ b/include/configs/microchip_mpfs_icicle.h
@@ -125,6 +125,9 @@
 #define CFG_EXTRA_ENV_SETTINGS \
 	"bootm_size=0x10000000\0" \
 	"scriptaddr=0x8e000000\0" \
+  "kernel_addr_r=0x8e000000\0" \
+  "ramdisk_addr_r=0x90000000\0" \
+  "fdt_addr_r=0x92000000\0" \
 	BOOTENV_DESIGN_OVERLAYS \
 	BOOTENV \
 
-- 
2.39.2

