diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 6ad59aeed5..36442cf0f5 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -85,7 +85,10 @@ dtb-$(CONFIG_ROCKCHIP_RK3288) += \
 dtb-$(CONFIG_ROCKCHIP_RK3368) += \
 	rk3368-sheep.dtb \
 	rk3368-geekbox.dtb \
-	rk3368-px5-evb.dtb \
+	rk3368-px5-evb.dtb
+
+dtb-$(CONFIG_ROCKCHIP_RK3588) += \
+	rk3588-orangepi-5-max.dtb
 
 dtb-$(CONFIG_ARCH_S5P4418) += \
 	s5p4418-nanopi2.dtb
